### ISSUE TITLE
make command failed

### ISSUE THREAD

**Issue by homedry** [hosseinmoein/DataFrame#222]
OS: Ubuntu 22.04.1 LTS
CMake: cmake version 3.22.1
Make: GNU Make 4.3
g++: g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

What I did:
`git clone https://github.com/hosseinmoein/DataFrame.git`
`cd DataFrame`
`mkdir Release`
`cd Release`
`cmake -DCMAKE_BUILD_TYPE=Release -DHMDF_BENCHMARKS=1 -DHMDF_EXAMPLES=1 -DHMDF_TESTING=1 ..`

And when I did:
`make`

> Consolidate compiler generated dependencies of target DataFrame
[ 14%] Built target DataFrame
[ 14%] Built target hmdf_copy_sample_data.json
[ 14%] Built target hmdf_copy_sample_data_dt_index.csv
[ 14%] Built target hmdf_copy_sample_data_string_index.csv
[ 14%] Built target hmdf_copy_sample_data.csv
Consolidate compiler generated dependencies of target dataframe_tester
[ 14%] Building CXX object test/CMakeFiles/dataframe_tester.dir/dataframe_tester.cc.o
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3803,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:317:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  317 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:319:1: error: template-id ‘_col_vector_push_back_<const char*, hmdf::DataFrame<I, H>::StlVecType<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >’ in declaration of primary template
  319 | _col_vector_push_back_<const char *, StlVecType<std::string>>(
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:340:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  340 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:342:1: error: template-id ‘_col_vector_push_back_<hmdf::DateTime, hmdf::DataFrame<I, H>::StlVecType<hmdf::DateTime> >’ in declaration of primary template
  342 | _col_vector_push_back_<DateTime, StlVecType<DateTime>>(
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:430:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  430 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:431:9: error: too few template-parameter-lists
  431 | struct  _IdxParserFunctor_<float>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:443:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  443 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:444:9: error: too few template-parameter-lists
  444 | struct  _IdxParserFunctor_<double>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:456:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  456 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:457:9: error: too few template-parameter-lists
  457 | struct  _IdxParserFunctor_<long double>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:469:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  469 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:470:9: error: too few template-parameter-lists
  470 | struct  _IdxParserFunctor_<int>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:482:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  482 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:483:9: error: too few template-parameter-lists
  483 | struct  _IdxParserFunctor_<long>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:495:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  495 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:496:9: error: too few template-parameter-lists
  496 | struct  _IdxParserFunctor_<long long>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:508:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  508 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:509:9: error: too few template-parameter-lists
  509 | struct  _IdxParserFunctor_<unsigned int>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:521:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  521 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:522:9: error: too few template-parameter-lists
  522 | struct  _IdxParserFunctor_<unsigned long>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:534:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  534 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:535:9: error: too few template-parameter-lists
  535 | struct  _IdxParserFunctor_<unsigned long long>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:547:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  547 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:548:9: error: too few template-parameter-lists
  548 | struct  _IdxParserFunctor_<std::string>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:564:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  564 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:565:9: error: too few template-parameter-lists
  565 | struct  _IdxParserFunctor_<DateTime>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:581:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  581 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:582:9: error: too few template-parameter-lists
  582 | struct  _IdxParserFunctor_<bool>  {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:641:10: error: explicit specialization in non-namespace scope ‘class hmdf::DataFrame<I, H>’
  641 | template<>
      |          ^
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:643:1: error: template-id ‘_generate_ts_index_<hmdf::DateTime>’ in declaration of primary template
  643 | _generate_ts_index_<DateTime>(StlVecType<DateTime> &index_vec,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3823,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc: In instantiation of ‘void hmdf::DataFrame<I, H>::read_csv_(std::istream&, bool) [with I = std::__cxx11::basic_string<char>; H = hmdf::HeteroVector<0>; std::istream = std::basic_istream<char>]’:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:778:9:   required from ‘bool hmdf::DataFrame<I, H>::read(S&, hmdf::io_format, bool) [with S = std::basic_istream<char>; I = std::__cxx11::basic_string<char>; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:763:23:   required from ‘bool hmdf::DataFrame<I, H>::read(const char*, hmdf::io_format, bool) [with I = std::__cxx11::basic_string<char>; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/test/dataframe_tester.cc:516:25:   required from here
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:330:44: error: no match for call to ‘(hmdf::DataFrame<std::__cxx11::basic_string<char>, hmdf::HeteroVector<0> >::_IdxParserFunctor_<std::__cxx11::basic_string<char> >) (hmdf::DataFrame<std::__cxx11::basic_string<char>, hmdf::HeteroVector<0> >::IndexVecType&, std::istream&)’
  330 |             _IdxParserFunctor_<IndexType>()(vec, stream);
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3803,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note: candidate: ‘void hmdf::DataFrame<I, H>::_IdxParserFunctor_<T>::operator()(hmdf::DataFrame<I, H>::StlVecType<T>&, std::istream&, hmdf::io_format) [with T = std::__cxx11::basic_string<char>; I = std::__cxx11::basic_string<char>; H = hmdf::HeteroVector<0>; hmdf::DataFrame<I, H>::StlVecType<T> = std::vector<std::__cxx11::basic_string<char> >; std::istream = std::basic_istream<char>]’
  425 |     void operator()(StlVecType<T> &, std::istream &, io_format)  {   }
      |          ^~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note:   candidate expects 3 arguments, 2 provided
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3823,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc: In instantiation of ‘void hmdf::DataFrame<I, H>::read_csv_(std::istream&, bool) [with I = hmdf::DateTime; H = hmdf::HeteroVector<0>; std::istream = std::basic_istream<char>]’:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:778:9:   required from ‘bool hmdf::DataFrame<I, H>::read(S&, hmdf::io_format, bool) [with S = std::basic_istream<char>; I = hmdf::DateTime; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:763:23:   required from ‘bool hmdf::DataFrame<I, H>::read(const char*, hmdf::io_format, bool) [with I = hmdf::DateTime; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/test/dataframe_tester.cc:531:24:   required from here
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:330:44: error: no match for call to ‘(hmdf::DataFrame<hmdf::DateTime, hmdf::HeteroVector<0> >::_IdxParserFunctor_<hmdf::DateTime>) (hmdf::DataFrame<hmdf::DateTime, hmdf::HeteroVector<0> >::IndexVecType&, std::istream&)’
  330 |             _IdxParserFunctor_<IndexType>()(vec, stream);
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3803,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note: candidate: ‘void hmdf::DataFrame<I, H>::_IdxParserFunctor_<T>::operator()(hmdf::DataFrame<I, H>::StlVecType<T>&, std::istream&, hmdf::io_format) [with T = hmdf::DateTime; I = hmdf::DateTime; H = hmdf::HeteroVector<0>; hmdf::DataFrame<I, H>::StlVecType<T> = std::vector<hmdf::DateTime, std::allocator<hmdf::DateTime> >; std::istream = std::basic_istream<char>]’
  425 |     void operator()(StlVecType<T> &, std::istream &, io_format)  {   }
      |          ^~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note:   candidate expects 3 arguments, 2 provided
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3823,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc: In instantiation of ‘void hmdf::DataFrame<I, H>::read_csv_(std::istream&, bool) [with I = long unsigned int; H = hmdf::HeteroVector<0>; std::istream = std::basic_istream<char>]’:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:778:9:   required from ‘bool hmdf::DataFrame<I, H>::read(S&, hmdf::io_format, bool) [with S = std::basic_istream<char>; I = long unsigned int; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:763:23:   required from ‘bool hmdf::DataFrame<I, H>::read(const char*, hmdf::io_format, bool) [with I = long unsigned int; H = hmdf::HeteroVector<0>]’
/media/Projects/DataFrame/test/dataframe_tester.cc:1242:17:   required from here
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_read.tcc:330:44: error: no match for call to ‘(hmdf::DataFrame<long unsigned int, hmdf::HeteroVector<0> >::_IdxParserFunctor_<long unsigned int>) (hmdf::DataFrame<long unsigned int, hmdf::HeteroVector<0> >::IndexVecType&, std::istream&)’
  330 |             _IdxParserFunctor_<IndexType>()(vec, stream);
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /media/Projects/DataFrame/include/DataFrame/DataFrame.h:3803,
                 from /media/Projects/DataFrame/test/dataframe_tester.cc:28:
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note: candidate: ‘void hmdf::DataFrame<I, H>::_IdxParserFunctor_<T>::operator()(hmdf::DataFrame<I, H>::StlVecType<T>&, std::istream&, hmdf::io_format) [with T = long unsigned int; I = long unsigned int; H = hmdf::HeteroVector<0>; hmdf::DataFrame<I, H>::StlVecType<T> = std::vector<long unsigned int, std::allocator<long unsigned int> >; std::istream = std::basic_istream<char>]’
  425 |     void operator()(StlVecType<T> &, std::istream &, io_format)  {   }
      |          ^~~~~~~~
/media/Projects/DataFrame/include/DataFrame/Internals/DataFrame_private_decl.h:425:10: note:   candidate expects 3 arguments, 2 provided
make[2]: *** [test/CMakeFiles/dataframe_tester.dir/build.make:76: test/CMakeFiles/dataframe_tester.dir/dataframe_tester.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:207: test/CMakeFiles/dataframe_tester.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

Please help me. Thank you in advance.

**Comment by hosseinmoein** [comment]
I cannot reproduce this. Are you sure your compiler is enabled for C++17 or higher?
The errors don't make sense, but that's typical C++

**Comment by hosseinmoein** [comment]
I just noticed my CMake file has this
```
cmake_minimum_required(VERSION 3.8)
```

**Comment by homedry** [comment]
my CMake file too. What version of CMake do you have?

> I cannot reproduce this. Are you sure your compiler is enabled for C++17 or higher? The errors don't make sense, but that's typical C++

this trivial program
```#include <iostream>
int main()
  { std::cout << __cplusplus << std::endl; }
```
prints output `201703` which means c++17.

**Comment by homedry** [comment]
> I cannot reproduce this. Are you sure your compiler is enabled for C++17 or higher? The errors don't make sense, but that's typical C++

this trivial program

int main()
  { std::cout << __cplusplus << std::endl; }
prints output 201703 which means c++17.

**Comment by homedry** [comment]
> I just noticed my CMake file has this
> 
> ```
> cmake_minimum_required(VERSION 3.8)
> ```


Should I downgrade cmake maybe?

but
`sudo apt list -a cmake`
prints 
```
cmake/jammy-updates,now 3.22.1-1ubuntu1.22.04.1 amd64 [installed]
cmake/jammy 3.22.1-1ubuntu1 amd64

cmake/jammy-updates 3.22.1-1ubuntu1.22.04.1 i386
cmake/jammy 3.22.1-1ubuntu1 i386
```

**Comment by homedry** [comment]
> > I cannot reproduce this. Are you sure your compiler is enabled for C++17 or higher? The errors don't make sense, but that's typical C++
> 
> this trivial program
> 
> ```
> int main()
>   { std::cout << __cplusplus << std::endl; }
> ```
> 
> prints output `201703` which means c++17.

What I tried:
I noticed in Release/CmakeCache.txt file, there is
`CMAKE_CXX_COMPILER:STRING=/usr/bin/c++`

So I added two line for setting compiler for cmake to use g++ not c++:
```
find_program(CMAKE_C_COMPILER NAMES $ENV{CC} gcc PATHS ENV PATH NO_DEFAULT_PATH)
find_program(CMAKE_CXX_COMPILER NAMES $ENV{CXX} g++ PATHS ENV PATH NO_DEFAULT_PATH)
```
But the same error reproduced.

**Comment by hosseinmoein** [comment]
Since I cannot reproduce this, my help would be very limited.
But as an experiment use my makefiles instead of CMake. They explicitly set `-std=c++20`.
Go to the `src` directory and do:
```
make -f Makefile.Linux.GCC64 clobber
make -f Makefile.Linux.GCC64
```

You may have to adjust the path to `g++` in the makefile

**Comment by hosseinmoein** [comment]
I was able to get the gcc and Ubuntu versions compatible with you and was able to reproduce the error. Apparently clang and VC++ do not produce this error. The conflict is supposed to have been resolved in C++20.
But I have a way of fixing it. Apparently, full specialization is not allowed within a template class, but partial specialization is allowed -- mind exploded.
It will take me over the weekend or maybe longer to find time to implement it.
